# Web Scraping

### What is it?

Web scraping is the art and practice of **retrieving information from a website**. Web scraping can be done manually and, in a sense, it's what we do every time we look up some information on the website or when we copy a table from a Wikipedia page and paste it into a spreadsheet.

That said, the most efficient way to scrape a website is through some kind of an automated script and we will explore this practice in Python using a well known scraping library called `BeautifulSoup`.

### How does it work?

1. The first step is to locate a website from which you want to extract some information that is embedded in it. It could be an e-Commerce, a news outlet, a weather site or any other website really.
2. You should then proceed to inspect and analyze the HTML source code of that web page to find the specific ids and tags that identify the information you're after.
3. The central part of a Python scraper is, of course, the script itself. You write an algorithm to instruct your program to use those tags and ids to retrieve the information you're after.
4. Finally, you manipulate your results into a data structure of your choice (csv, json, ...) that will allow you to interact with and use your scraped data.

<img src="images/web-scraping.png" width="600">

### Is it legal?

In general terms, web scraping is not illegal. More specifically, according to Wikipedia: *the legality of web scraping varies across the world. In general, web scraping may be against the terms of use of some websites, but the enforceability of these terms is unclear.*

It is generally accepted that, if a website makes its data public, then scraping it should be legal. This is reinforced by legal cases such as "LinkedIn vs hiQ Labs". You can read more on [Wikipedia](https://en.wikipedia.org/wiki/Web_scraping#Legal_issues) and [online](https://www.parsehub.com/blog/web-scraping-legal/).

### A first simple example

In [1]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup

We will start with a [simple page](https://www.scrapethissite.com/pages/simple/), click on the link and familiarise yourself with its contents; it contains a list of country names from the world and some information for each one of them.

Before moving on, inspect the underlying HTML code: right click on an element of the page (for example the name "Andorra") and select "Inspect". This will open the Inspect Element window in your Chrome browser, which allows you to see the code that generates the page you're looking at.

<img src="images/inspect-element.png" width="600">

In Python, let's start by sending an HTTP GET request to retrieve the HTML content of a desired web page and print to screen the first 1000 characters:

In [2]:
url = "https://www.scrapethissite.com/pages/simple/"
page = requests.get(url)

print(page.text[0:1000])

<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Countries of the World: A Simple Example | Scrape This Site | A public sandbox for learning web scraping</title>
    <link rel="icon" type="image/png" href="/static/images/scraper-icon.png" />

    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="A single page that lists information about all the countries in the world. Good for those just get started with web scraping.">

    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap.min.css" rel="stylesheet" integrity="sha256-MfvZlkHCEqatNoGiOXveE8FIwMzZg4W85qfrfIFBfYc= sha512-dTfge/zgoMYpP7QbHy4gWMEGsbsdZeCXz7irItjcC3sPUFtf0kuFbDz/ixG7ArTxmDjLXDmezHubeNikyKGVyQ==" crossorigin="anonymous">
    <link href='https://fonts.googleapis.com/css?family=Lato:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="/static/css/styles.css">

    
<meta name=

In [3]:
type(page.text)

str

Then we create a Beautiful Soup object named `soup` that takes `page.content` as an input (this is the HTML content we just printed to screen).

*Note: it is better to pass page.content instead of page.text to avoid issues with character encoding.*

In [4]:
soup = BeautifulSoup(page.content, "html.parser")

In [5]:
type(soup)

bs4.BeautifulSoup

If you go back to the **Inspect Element** window, you'll notice that if you hover over an element in the windown on the right, it will highlight on the webpage to the left. We want to retrieve all the country names in the list and the `<div id="page">` element contains all of the results that we're after, so we apply the `.find()` method to the `soup` object in order to find that element via its `id="page"` parameter.

<img src="images/div id=page.png" width="600">

In [6]:
results = soup.find(id="page")

In [7]:
results

<div id="page">
<section id="countries">
<div class="container">
<div class="row">
<div class="col-md-12">
<h1>
                            Countries of the World: A Simple Example
                            <small>250 items</small>
</h1>
<hr/>
</div>
</div>
<div class="row">
<div class="col-md-12">
<p class="lead">
                            A single page that lists information about all the countries in the world. Good for those just get started with web scraping.
                            Practice looking for patterns in the HTML that will allow you to extract information about each country. Then, build a simple web scraper that makes a request to this page, parses the HTML and prints out each country's name.
                        </p>
<hr/>
</div>
</div>
<div class="row">
<div class="col-md-6">
<p>
<i class="glyphicon glyphicon-education"></i> There are <a href="/lessons/">4 video lessons</a> that show you how to scrape this page.
                        </p>
<hr/>
</div>
<di

As we were saying, inside the `results` object there are all the country names we're interested in, but we still need to identify them and to do that we go back to the Inspect Element window and notice that each country block has its name and all related information stored inside a `div` element with `class="col-md-4 country"`. Therefore, we apply the `.find_all()` method to the `results` object in order to get the information about each country.

In [8]:
countries = results.find_all("div", class_="col-md-4 country")

In [9]:
type(countries)

bs4.element.ResultSet

In [10]:
len(countries)

250

In [11]:
countries[0]

<div class="col-md-4 country">
<h3 class="country-name">
<i class="flag-icon flag-icon-ad"></i>
                            Andorra
                        </h3>
<div class="country-info">
<strong>Capital:</strong> <span class="country-capital">Andorra la Vella</span><br/>
<strong>Population:</strong> <span class="country-population">84000</span><br/>
<strong>Area (km<sup>2</sup>):</strong> <span class="country-area">468.0</span><br/>
</div>
</div>

In [12]:
type(countries[0])

bs4.element.Tag

Now, if you look closely to the above output, you'll notice that each country name is enclosed in a `h3` tag with `class="country-name"`.

In [13]:
type(countries)

bs4.element.ResultSet

In [14]:
len(countries)

250

The output object `countries` is an iterable object of class `bs4.element.ResultSet` containing 250 elements, so we can use a for loop to print each element in the `h3` tag.

In [15]:
countries[0:3]

[<div class="col-md-4 country">
 <h3 class="country-name">
 <i class="flag-icon flag-icon-ad"></i>
                             Andorra
                         </h3>
 <div class="country-info">
 <strong>Capital:</strong> <span class="country-capital">Andorra la Vella</span><br/>
 <strong>Population:</strong> <span class="country-population">84000</span><br/>
 <strong>Area (km<sup>2</sup>):</strong> <span class="country-area">468.0</span><br/>
 </div>
 </div>,
 <div class="col-md-4 country">
 <h3 class="country-name">
 <i class="flag-icon flag-icon-ae"></i>
                             United Arab Emirates
                         </h3>
 <div class="country-info">
 <strong>Capital:</strong> <span class="country-capital">Abu Dhabi</span><br/>
 <strong>Population:</strong> <span class="country-population">4975593</span><br/>
 <strong>Area (km<sup>2</sup>):</strong> <span class="country-area">82880.0</span><br/>
 </div>
 </div>,
 <div class="col-md-4 country">
 <h3 class="country-name">
 

In [16]:
for c in countries[0:3]:
    print(c.find("h3", class_="country-name"))

<h3 class="country-name">
<i class="flag-icon flag-icon-ad"></i>
                            Andorra
                        </h3>
<h3 class="country-name">
<i class="flag-icon flag-icon-ae"></i>
                            United Arab Emirates
                        </h3>
<h3 class="country-name">
<i class="flag-icon flag-icon-af"></i>
                            Afghanistan
                        </h3>


As you can see, this prints everything, including the tag itself; in order **to include just the text**, we can add the `.text` attribute.

In [17]:
for c in countries[0:3]:
    print(c.find("h3", class_="country-name").text)



                            Andorra
                        


                            United Arab Emirates
                        


                            Afghanistan
                        


Now, instead of printing the output to screen, it's better to append each new element to a list in order to store those information for a later use.

In [18]:
listy = []
for c in countries:
    listy.append(c.find("h3", class_="country-name").text)

In [19]:
listy[0:5]

['\n\n                            Andorra\n                        ',
 '\n\n                            United Arab Emirates\n                        ',
 '\n\n                            Afghanistan\n                        ',
 '\n\n                            Antigua and Barbuda\n                        ',
 '\n\n                            Anguilla\n                        ']

Since there are some "spaces" and "return" characters in the strings, we use the `.strip()` method to remove those from the output.

In [20]:
type(listy[0])

str

In [21]:
names = []
for c in countries:
    names.append(c.find("h3", {"class":"country-name"}).text.strip())

In [22]:
names

['Andorra',
 'United Arab Emirates',
 'Afghanistan',
 'Antigua and Barbuda',
 'Anguilla',
 'Albania',
 'Armenia',
 'Angola',
 'Antarctica',
 'Argentina',
 'American Samoa',
 'Austria',
 'Australia',
 'Aruba',
 'Åland',
 'Azerbaijan',
 'Bosnia and Herzegovina',
 'Barbados',
 'Bangladesh',
 'Belgium',
 'Burkina Faso',
 'Bulgaria',
 'Bahrain',
 'Burundi',
 'Benin',
 'Saint Barthélemy',
 'Bermuda',
 'Brunei',
 'Bolivia',
 'Bonaire',
 'Brazil',
 'Bahamas',
 'Bhutan',
 'Bouvet Island',
 'Botswana',
 'Belarus',
 'Belize',
 'Canada',
 'Cocos [Keeling] Islands',
 'Democratic Republic of the Congo',
 'Central African Republic',
 'Republic of the Congo',
 'Switzerland',
 'Ivory Coast',
 'Cook Islands',
 'Chile',
 'Cameroon',
 'China',
 'Colombia',
 'Costa Rica',
 'Cuba',
 'Cape Verde',
 'Curacao',
 'Christmas Island',
 'Cyprus',
 'Czech Republic',
 'Germany',
 'Djibouti',
 'Denmark',
 'Dominica',
 'Dominican Republic',
 'Algeria',
 'Ecuador',
 'Estonia',
 'Egypt',
 'Western Sahara',
 'Eritrea',
 

In [23]:
countries[0]

<div class="col-md-4 country">
<h3 class="country-name">
<i class="flag-icon flag-icon-ad"></i>
                            Andorra
                        </h3>
<div class="country-info">
<strong>Capital:</strong> <span class="country-capital">Andorra la Vella</span><br/>
<strong>Population:</strong> <span class="country-population">84000</span><br/>
<strong>Area (km<sup>2</sup>):</strong> <span class="country-area">468.0</span><br/>
</div>
</div>

In [24]:
int(countries[0].find("span", class_="country-population").text)

84000

In [25]:
names[0:5]

['Andorra',
 'United Arab Emirates',
 'Afghanistan',
 'Antigua and Barbuda',
 'Anguilla']

That's it, we successfully created a list including all the country names from our initial web page.

### A slightly more difficult example

At [this link](https://www.rottentomatoes.com/browse/movies_at_home/affiliates:netflix~sort:critic_highest) you will find a list of the top movies (according to the website's Tomatometer score) currently available on Netflix. We want to retrieve some information (such as the movie title) for all the movies available in the page and save such information in a pandas data frame.

We begin by sending an HTTP GET request to retrieve the HTML page and save its contents to a `BeautifulSoup` object.

*Notice the `headers={'Accept-Language': "lang=en-US"}` additional parameter, which ensures that the contents are always retrieved in english, no matter what the client language is.*

In [26]:
url = 'https://www.rottentomatoes.com/browse/movies_at_home/affiliates:netflix~sort:critic_highest'
page = requests.get(url, headers={'Accept-Language': "lang=en-US"})

In [27]:
page.status_code

200

In [28]:
soup = BeautifulSoup(page.content, "html.parser")

Using the Element Inspector tool, we can see that each movie title belongs to the `div` tag and is identified by the `class="js-tile-link"` parameter. Using the `.find_all()` method, we can find all those occurrences; we then save them to an object called `movies`.

In [29]:
movies = soup.find_all('div', class_='flex-container')

In [30]:
type(movies)

bs4.element.ResultSet

The `movies` object is a BeautifulSoup `ResultSet` object, which is an iterable and contains all the movie titles we're interested in. Let's explore the first element in the result set:

In [31]:
movies[0]

<div class="flex-container" data-ems-id="353540c1-20f4-304e-b146-f769fa105a5d">
<media-info-tile>
<poster-tile data-qa="discovery-media-list-item" media-url="/m/a_love_song_for_latasha" size-tshirt="none,xxx-large" slot="poster">
<rt-img alt="A Love Song for Latasha" class="posterImage" loading="lazy" slot="thumbnail" src="https://resizing.flixster.com/k7WNrT9Gtgyh_W9XX99OR6RGXRE=/206x305/v2/https://resizing.flixster.com/-XZAfHZM39UwaGJIFWKAE8fS0ak=/v3/t/assets/p18702247_p_v10_aa.jpg"></rt-img>
</poster-tile>
<a class="js-tile-link" data-ems-id="353540c1-20f4-304e-b146-f769fa105a5d" data-qa="discovery-media-list-item-caption" data-track="scores" href="/m/a_love_song_for_latasha" slot="meta-data">
<score-pairs-deprecated>
<score-icon-critics certified="false" sentiment="positive" size="1" slot="criticsScoreIcon"></score-icon-critics>
<rt-text context="label" size="1" slot="criticsScore"> 100%</rt-text>
<score-icon-audience certified="false" sentiment="positive" size="1" slot="audienceSc

We can see that each element of the `movies` object contains a series of `rt-text` tags, from which we can extract their text content. Let's use the `.find_all()` method to extract each one of these from the first element in the `movies` object:

*Note: in HTML, an `rt-text` is a custom HTML element, specific to the Rotten Tomatoes website. it is recognizable by the mandatory hyphen in its name, which is the W3C convention for custom elements.*

In [32]:
movies[0].find('rt-text')

<rt-text context="label" size="1" slot="criticsScore"> 100%</rt-text>

We're interested in the second `<rt-text>` tag, as that's the one that contains the movie title, so let's use the `[2]` index to isolate it; furthermore, as we now know, to extract its text content we need to use the `.text` attribute:

In [33]:
movies[0].find_all('rt-text')

[<rt-text context="label" size="1" slot="criticsScore"> 100%</rt-text>,
 <rt-text context="label" size="1" slot="audienceScore"> 79%</rt-text>,
 <rt-text data-qa="discovery-media-list-item-title" size="0.875">
           A Love Song for Latasha
         </rt-text>,
 <rt-text data-qa="discovery-media-list-item-start-date" size="0.75">
             Streaming Sep 24, 2020
           </rt-text>]

In [34]:
movies[0].find_all('rt-text')[2]

<rt-text data-qa="discovery-media-list-item-title" size="0.875">
          A Love Song for Latasha
        </rt-text>

In [35]:
movies[0].find_all('rt-text')[2].text

'\n          A Love Song for Latasha\n        '

Applying the `.strip()` method to this result, we finally get a cleaned up version of the movie title:

In [36]:
movies[0].find_all('rt-text')[2].text.strip()

'A Love Song for Latasha'

In [37]:
movies[0].find('rt-text', {"context":"label", "slot":"audienceScore"}).text.strip()
# movies[0].find('rt-text', slot="audienceScore").text.strip()

'79%'

What we have done so far is extracting the movie title from the first element in the `movies` object. In order to **extract all the movie titles**, let's cycle through the `movies` object, extract each movie title and save it to a new list that we'll name `movie_names`:

In [38]:
movie_names = []
for m in movies:
    movie_names.append(m.find_all('rt-text')[2].text.strip())

In [39]:
movie_names[0:5]

['A Love Song for Latasha',
 'A Nonsense Christmas with Sabrina Carpenter',
 'A Random Desi Romance',
 'A Secret Love',
 'AKA']

Now, using the same method, we proceed to extract the streaming start date (which is stored in the third `<rt-text>` element):

In [40]:
movie_dates = []
for m in movies:
    movie_dates.append(m.find_all('rt-text')[3].text.strip())

In [41]:
movie_dates[0:5]

['Streaming Sep 24, 2020',
 'Streaming Dec 6, 2024',
 'Streaming Dec 16, 2015',
 'Streaming Apr 29, 2020',
 'Streaming Apr 28, 2023']

At this point, we'd like to retrieve the audience and critic scores, which are stored in the `<score-pairs-deprecated>` tag; let's use the `.find()` method to isolate this specific tag:

In [42]:
movies[0]

<div class="flex-container" data-ems-id="353540c1-20f4-304e-b146-f769fa105a5d">
<media-info-tile>
<poster-tile data-qa="discovery-media-list-item" media-url="/m/a_love_song_for_latasha" size-tshirt="none,xxx-large" slot="poster">
<rt-img alt="A Love Song for Latasha" class="posterImage" loading="lazy" slot="thumbnail" src="https://resizing.flixster.com/k7WNrT9Gtgyh_W9XX99OR6RGXRE=/206x305/v2/https://resizing.flixster.com/-XZAfHZM39UwaGJIFWKAE8fS0ak=/v3/t/assets/p18702247_p_v10_aa.jpg"></rt-img>
</poster-tile>
<a class="js-tile-link" data-ems-id="353540c1-20f4-304e-b146-f769fa105a5d" data-qa="discovery-media-list-item-caption" data-track="scores" href="/m/a_love_song_for_latasha" slot="meta-data">
<score-pairs-deprecated>
<score-icon-critics certified="false" sentiment="positive" size="1" slot="criticsScoreIcon"></score-icon-critics>
<rt-text context="label" size="1" slot="criticsScore"> 100%</rt-text>
<score-icon-audience certified="false" sentiment="positive" size="1" slot="audienceSc

In [43]:
movies[0].find('score-pairs-deprecated')

<score-pairs-deprecated>
<score-icon-critics certified="false" sentiment="positive" size="1" slot="criticsScoreIcon"></score-icon-critics>
<rt-text context="label" size="1" slot="criticsScore"> 100%</rt-text>
<score-icon-audience certified="false" sentiment="positive" size="1" slot="audienceScoreIcon"></score-icon-audience>
<rt-text context="label" size="1" slot="audienceScore"> 79%</rt-text>
</score-pairs-deprecated>

You may have noticed that the `score-pairs-deprecated` tag has an attribute called `audienceScore` and one called `criticsScore`; using these two attributes, we'll be able to to extract the corresponding scores that we're interested in.

To access this information, we'll use the following combination `.find('rt-text', {'slot': 'ATTR_NAME'}).get_text(strip=True)`, where:

- the `.find()` method is used to locate the `rt-text` tags with specific `slot` attributes (`audienceScore` and `criticsScore`) ,
- the `get_text(strip=True)` method is used to get the text content without any leading or trailing whitespace.

In [44]:
movies[0].find('score-pairs-deprecated').find('rt-text', {'slot': 'audienceScore'}).get_text(strip=True)

'79%'

In [45]:
movies[0].find('score-pairs-deprecated').find('rt-text', {'slot': 'criticsScore'}).get_text(strip=True)

'100%'

Let's loop through all elements in the `movies` object to extract the audience and critics scores:

In [46]:
movie_scores_audience = []
movie_scores_critics = []
for m in movies:
    movie_scores_audience.append(m.find('score-pairs-deprecated').find('rt-text', {'slot': 'audienceScore'}).get_text(strip=True))
    movie_scores_critics.append(m.find('score-pairs-deprecated').find('rt-text', {'slot': 'criticsScore'}).get_text(strip=True))

In [47]:
movie_scores_audience[0:5]

['79%', '73%', '43%', '88%', '73%']

In [48]:
movie_scores_critics[0:5]

['100%', '100%', '100%', '100%', '100%']

Finally, let's combine together the four lists that we have created into a single DataFrame named `df_movies`:

In [49]:
df_movies = pd.DataFrame(
    {'movie_name': movie_names,
     'streaming_date': movie_dates,
     'audience_score': movie_scores_audience,
     'critics_score': movie_scores_critics
    })

In [50]:
df_movies

,movie_name,streaming_date,audience_score,critics_score
0,A Love Song for Latasha,"Streaming Sep 24, 2020",79%,100%
1,A Nonsense Christmas with Sabrina Carpenter,"Streaming Dec 6, 2024",73%,100%
2,A Random Desi Romance,"Streaming Dec 16, 2015",43%,100%
3,A Secret Love,"Streaming Apr 29, 2020",88%,100%
4,AKA,"Streaming Apr 28, 2023",73%,100%
5,Ajeeb Daastaans,"Streaming Apr 16, 2021",66%,100%
6,Ali Wong: Baby Cobra,"Streaming May 6, 2016",74%,100%
7,Ali Wong: Hard Knock Wife,"Streaming May 13, 2018",68%,100%
8,Alive and Kicking,"Streaming Apr 7, 2017",94%,100%
9,Amy Schumer: Emergency Contact,"Streaming Jun 13, 2023",14%,100%


In [51]:
df_movies.shape

(28, 4)

One last thing that we can do is to clean the data frame by:

- formatting the `streaming_date` column so that it is interpreted as a date type;
- converting missing scores to a numpy NaN value (otherwise you won't be able to cast it as a float);
- remove the `%` symbol from the critics and audience scores;
- cast the two scores as integers;
- sort the data frame by the `critics_score` and then by the `audience_score`.

In [52]:
# format as a date data type
df_movies['streaming_date'] = pd.to_datetime(df_movies['streaming_date'].str.replace('Streaming ', ''))

In [54]:
# convert missing scores to a numpy NaN value
df_movies.loc[df_movies['audience_score']=='', 'audience_score'] = np.nan
df_movies.loc[df_movies['critics_score']=='', 'critics_score'] = np.nan

In [55]:
df_movies['audience_score'] = df_movies['audience_score'].str.replace('%', '')
df_movies['critics_score'] = df_movies['critics_score'].str.replace('%', '')

In [56]:
# convert the scores to floats
df_movies = df_movies.astype({'audience_score':'float', 'critics_score':'float'})

In [57]:
# check data types
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28 entries, 0 to 27
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   movie_name      28 non-null     object        
 1   streaming_date  28 non-null     datetime64[ns]
 2   audience_score  25 non-null     float64       
 3   critics_score   27 non-null     float64       
dtypes: datetime64[ns](1), float64(2), object(1)
memory usage: 1.0+ KB


In [58]:
# sorting data frame
df_movies.sort_values(['critics_score', 'audience_score'], ascending = False, inplace = True)

In [59]:
df_movies

,movie_name,streaming_date,audience_score,critics_score
16,Blue Giant,2024-04-30,97.0,100.0
8,Alive and Kicking,2017-04-07,94.0,100.0
12,Athlete A,2020-06-25,92.0,100.0
27,Cuba and the Cameraman,2017-11-24,92.0,100.0
17,Bo Burnham: Make Happy,2016-06-03,91.0,100.0
21,Chasing Coral,2017-07-15,89.0,100.0
3,A Secret Love,2020-04-29,88.0,100.0
10,Apollo 13: Survival,2024-09-05,88.0,100.0
0,A Love Song for Latasha,2020-09-24,79.0,100.0
13,Audible,2021-07-01,79.0,100.0


### It's your turn now

Find the easiest way to download the information of only the movies on the second [page](https://www.rottentomatoes.com/browse/movies_at_home/affiliates:netflix~sort:critic_highest?page=2). Append the newly retrieved movie titles to the df_movies data frame.

In [61]:
# get page source code
url = 'https://www.rottentomatoes.com/browse/movies_at_home/affiliates:netflix~sort:critic_highest?page=2'
page2 = requests.get(url, headers={'Accept-Language': "lang=en-US"})

page2.status_code

soup = BeautifulSoup(page2.content, "html.parser")

In [69]:
# retrieve movies information
movies = soup.find_all('div', class_='flex-container')
movies = movies[28:]  # Skip the first element which is not a movie

print(len(movies))

movie_names_page2 = []
for m in movies:
    movie_names_page2.append(m.find_all('rt-text')[2].text.strip())

print(movie_names_page2[0:5])

movie_dates_page2 = []
for m in movies:
    movie_dates_page2.append(m.find_all('rt-text')[3].text.strip())

print(movie_dates_page2[0:5])

movie_scores_audience_page2 = []
movie_scores_critics_page2 = []
for m in movies:
    movie_scores_audience_page2.append(m.find('score-pairs-deprecated').find('rt-text', {'slot': 'audienceScore'}).get_text(strip=True))
    movie_scores_critics_page2.append(m.find('score-pairs-deprecated').find('rt-text', {'slot': 'criticsScore'}).get_text(strip=True))

print(movie_scores_audience_page2[0:5])
print(movie_scores_critics_page2[0:5])

28
['Cunk on Life', 'Daughters', 'Descendant', 'End Game', 'Evelyn']
['Streaming Jan 2, 2025', 'Streaming Aug 14, 2024', 'Streaming Oct 21, 2022', 'Streaming May 4, 2018', 'Streaming Sep 10, 2019']
['81%', '91%', '64%', '75%', '75%']
['100%', '100%', '100%', '100%', '100%']


In [70]:
# create new data frame
df_movies_page2 = pd.DataFrame({
    'movie_name': movie_names_page2,
    'streaming_date': movie_dates_page2,
    'audience_score': movie_scores_audience_page2,
    'critics_score': movie_scores_critics_page2
})

df_movies_page2.head()

,movie_name,streaming_date,audience_score,critics_score
0,Cunk on Life,"Streaming Jan 2, 2025",81%,100%
1,Daughters,"Streaming Aug 14, 2024",91%,100%
2,Descendant,"Streaming Oct 21, 2022",64%,100%
3,End Game,"Streaming May 4, 2018",75%,100%
4,Evelyn,"Streaming Sep 10, 2019",75%,100%


In [71]:
df_movies_page2['streaming_date'] = pd.to_datetime(df_movies_page2['streaming_date'].str.replace('Streaming ', ''))

df_movies_page2.loc[df_movies_page2['audience_score']=='', 'audience_score'] = np.nan
df_movies_page2.loc[df_movies_page2['critics_score']=='', 'critics_score'] = np.nan

df_movies_page2['audience_score'] = df_movies_page2['audience_score'].str.replace('%', '')
df_movies_page2['critics_score'] = df_movies_page2['critics_score'].str.replace('%', '')

df_movies_page2 = df_movies_page2.astype({'audience_score':'float', 'critics_score':'float'})

In [72]:
# check data frame shape
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 28 entries, 16 to 25
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   movie_name      28 non-null     object        
 1   streaming_date  28 non-null     datetime64[ns]
 2   audience_score  25 non-null     float64       
 3   critics_score   27 non-null     float64       
dtypes: datetime64[ns](1), float64(2), object(1)
memory usage: 1.1+ KB


In [75]:
# append new dataframe (df_movies_page2) to old one (df_movies)
df = pd.concat([df_movies, df_movies_page2], ignore_index=True)

df.head()

,movie_name,streaming_date,audience_score,critics_score
0,Blue Giant,2024-04-30,97.0,100.0
1,Alive and Kicking,2017-04-07,94.0,100.0
2,Athlete A,2020-06-25,92.0,100.0
3,Cuba and the Cameraman,2017-11-24,92.0,100.0
4,Bo Burnham: Make Happy,2016-06-03,91.0,100.0


In [76]:
# sort data frame
df.sort_values(['critics_score', 'audience_score'], ascending = False, inplace = True)

In [77]:
df

,movie_name,streaming_date,audience_score,critics_score
0,Blue Giant,2024-04-30,97.0,100.0
1,Alive and Kicking,2017-04-07,94.0,100.0
41,If Anything Happens I Love You,2020-11-20,93.0,100.0
2,Athlete A,2020-06-25,92.0,100.0
3,Cuba and the Cameraman,2017-11-24,92.0,100.0
4,Bo Burnham: Make Happy,2016-06-03,91.0,100.0
29,Daughters,2024-08-14,91.0,100.0
45,John Mulaney: Kid Gorgeous at Radio City,2018-05-01,90.0,100.0
5,Chasing Coral,2017-07-15,89.0,100.0
6,A Secret Love,2020-04-29,88.0,100.0
